### DISCUSSION TITLE
What am I doing wrong?

### DISCUSSION THREAD

**Discussion by jonnypoko** [frntc/SIDKick-pico#39] [Q&A] [answered]
Made up a few of these beautiful boards and tried with two official RPi Pico H - both work fine (very excited) Then I tried with three RPi Pico RP240 clones (4M and 16M veresions, non LED) and none of them work - no sound and no config menu? all three of these boards work with the blink.uf2? what could I be doing wrong, I'm afraid my knowlege of RPi ends about here? Awesome project!! what is the difference between the LED and RGB uf2 files (where do I find that info? - sorry pretty new to github too)

**Comment by frntc** [comment]
I hope you did not try the purple clones? Because they have another pinout and won't work... (flash size 4M, 16M, ... does not matter)

LED = firmware that lets the status LED blink, RGB = firmware for clones with RGB LED on the PCB.

**Reply by jonnypoko** [reply]
purple as in pcb colour? mine are black? I have ordered a couple of the RGB LED ones as well and they should arrive soon.

**Reply by frntc** [reply, marked as answer]
Color aside :-) -- is the pinout the same as for the original Pi Pico?

Several people experience bad build quality for the cheap clones (wrongly soldered parts, solder bridges etc), so the best thing is to also try a "blink" example to see if it works at all

**Reply by jonnypoko** [reply]
Blink works on all of them, but you are correct the pinout is different??? why would they sell clones with different pinouts, it's a black board (both the 4M and 16M versions) which has some random pin config. I should have checked this I guess, but it is hard to decipher from the various order links on AliEx. So a **WARNING TO ALL** not all clones are created equally - even before you get to build quality issues. Is there a differnet RPIPico board they are trying to emulate with this pinout do you know? Anyway, the boards with the RGB on seem to be a match (arrived today) so I will try those tonight. Final Question - what are the minimum pins required to hook up the SidKickPico and the RPi board?

**Reply by jonnypoko** [reply]
I just wanted to add, thanks for the help and that I have never been more excited about a new project than after watching the "more fun fixing it" youtube video on the SIDKickPico - I ordered the boards that night...👍

**Reply by frntc** [reply]
you mean which GPIOs from the Pico (clones) are needed for the SKpico? all that are present on the original Pico :-)

**Reply by jonnypoko** [reply]
next phase of my journey from planet stupid - I got the new boards with the RGB LED on and the first one i have soldered works fine - sound, menu - but no RGB? I preesume the sidkick pico board is outputting the WS2812 to GPIO23 on pin 37?

**Reply by frntc** [reply]
The clones with the RGB-LED connect it directly to the RP2040, you only have to take the right firmware ... and: one some of these boards there is an option little jumpersolder directly above the RGB-LED which you have to close.

**Reply by jonnypoko** [reply]
Cheers, sorted - it would help those of us with numb skulls if the 'somewhat more reputable' supplier of the RGB boards I got (they have a website printed on the back of the Pi where you can download schematics pinouts and pictures of the boards, yay) had markings on the board that matched the schematic - jumper labeled as R7 on the silkscreen, however R7 on the schematic is at the other end of the board - oh well.. thanks for all the answers! and the cool projects!

**Reply by rsquared68** [reply]
I found that using the relatively common black-colored YD-RP2040 boards, firmware after v0.14 is not usable for me.  The version string is corrupted, on first entry the configure menu is corrupted, can't save/load configurations, and the prg slots menu crashes. I've only done a little investigation so far, but I think it's because the black YD-2040 boards have 4MB instead of the standard 2MB.  I was able to rebuild with pico.h modified so that the flash size of the target platform is 4 * 1024 * 1024 and all of those features now seem to work with this board after that change.  I have not tested it extensively but it doesn't crash or display a lot of garbage characters.  However, my build does not make sound...either because I had to comment out some lost stuff (RESET, bRESET, resetEverything() etc) to complete the build or because I re-mapped the GPIOs incorrectly.  In the .zip source I downloaded the GPIO mapping did not seem correct for the v0.1 PCB (all I have) based on me poking around with a multimeter lol.  So if you are able to build, try changing the flash size for the black boards.

**Reply by frntc** [reply]
As long as the pinout is the same, these YD-RP2040 boards should work (do you have a link?). The flash size does not make any difference, no need to recompile the firmware. (I'm mostly using 16MB boards.)

**Comment by rsquared68** [comment]
I think the pinout is the same except for 3V3_EN, but I have no real Picos.  https://github.com/initdc/YD-RP2040/blob/master/YD-2040-PIN.png

I’m super interested in your projects because for a long time I wanted to experiment with C64 bus interfacing—your stuff is great!  If you want I can send you detailed documentation of what I know so far, but using the v0.1 PCB in a PAL 326298 and Kawari clock I find the following (I don’t think this is the problem because friend reproduced similar behavior in a stock NTSC machine with the uf2 files in the 0.202 release):


  1.  V0.14 .uf2 gitHub release works perfectly



  1.  V0.20+ .uf2 gitHub release makes sound, but version string is trash, on initial entry configurator screen has some other garbage, save/load config doesn’t result in any changes, prg launcher screen is filled with garbage and crashes C64.  However, it emulates some SID configuration and makes sound OK.  It just seems when interacting with flash there is a problem.  I tried several different .uf2 files from the release for both PCB versions.  They all behave identically in this respect.


  1.  Because flash seemed implicated I wanted to investigate, and I tried to build from the .zip archive in the release—I did not clone the repo bc I am experimenting with sandboxing VSCode.  Anyway, it would not build initially because RESET and bRESET are undefined but used, and extern resetEverything() is nowhere in that zip archive.  On PCB v0.1 SID reset is connected to RUN so there is no way to define RESET or bRESET mask so I just commented out the offending stuff. I found resetEverything() in the firmware v0.10 archive.  Builds OK but does nothing in the C64.


  1.  I next looked at the GPIO mapping to start to understand the interface—it was clear to me that there was no way the code could work because the code suggested that RESET was connected to a GPIO but in the circuit it was not.  I discovered that in the .zip source release, the code had (for example) phi2 on GPIO 12 when it is actually on GPIO 9 (passing through the lower FET buffer) on the PCB I have.   Someone gave me the boards and said they were from PCBWay; that’s all I know about the origin.  It has the same silkscreen etc as the photos in your repo but maybe this design is somehow different?  After probing around with a multimeter and logic analyzer, I had to redefine all of the GPIO mappings except for D0-7 and the LEDs.


  1.  After redefining the GPIO mappings the config menu started to work, but had the same corrupt version string and graphics, but also no sound.  I increased the flash size in pico.h (actually created a new board definition) and rebuilt the pico-sdk. After increasing the size, all of the flash-related problems went away, no corruption of version or graphics, prg launcher menu works, etc.


  1.  Sound emulation (DAC and PWM) and the onboard LED do not work with my modified build yet.  It’s weird because the config/launcher works, and even the paddles work.  So either reSID isn’t running properly, or possibly RW is somehow stuck and the CPU can only read from the SIDkick.  It’s possible that reSID is somehow not happy because of the workarounds I put in for RESET and bRESET, but the LED (not the RGB LED) doesn’t show activity either so maybe something else is stuck in runEmulation.  The bus handling otherwise seems ok.

When I rebuilt with the larger flash size, I only rebuilt the SDK.  It occurs to me that the memmap copy to ram or whatever also has the wrong size hard-coded in it so maybe I should adjust.

If you have any specific requests for things to try to further debug or for testing on weird machines let me know.  Currently I have one mostly stock 64C from Germany (has a TOLB clock in it though), a US locale breadbin with a VIC-II^2 switcher ntsc and pal, and a US locale breadbin with a VIC-II Kawari.  The friend that gave me the unpopulated PCBs probably has 30 more 😊  Once I get this to work I’m curious to see how it behaves with my own long-neglected project (https://github.com/rsquared68/schreckenstein64/releases/tag/rc1.0) – I added sound at the last minute and since I don’t fully understand hard-restart the gating is done with extreme violence at the moment.  It also relies on OSC3 for fast random number generation, and some emulations really mess up the game progress if this isn’t working well.

Best regards


________________________________

**Reply by frntc** [reply]
Just a quick answer for today: the Pico-clone pinout is fine, I'm using the same ones here. The problem you describe is certainly not the flash size, as I'm using different flash sizes. The size you can change in the source is just to amount of flash _reserved_ for PRGs. And yes, RESET is connect to a GPIO, depending on the hardware revision... (also there are more hw revisions existing than those that I've made available so far ;-)).

The most likely reason (if I gathered the information right) for your problems (it plays, but config menu doesn't work) is that the bus communication is unstable -- I know that I need to touch the config tool because of some NTSC machines, and the Kawari is said to cause different timings on the bus than the original VIC (but I don't have one myself).

My recommendation is: first experiment with the timings, you can do that with SKpicopatch.exe. For everything else: we should maybe take this offline, you'll find my mail address in the source.

**Comment by rsquared68** [comment]
I'll follow up--I found by building from the source that my specific problem with the released .uf2 was related to flash memory speed and everything else was a red herring.   The memcpy from flash commands in readConfiguration() didn't work properly for me when the clock is 300MHz—maybe the flash is slower on this clone board.  The code listing in the source .zip archive does SET_CLOCK_FAST *after* the memcpy, but nothing performs a SET_CLOCK_125MHZ before.  Maybe the slowdown macro had been in the code at some point and got deleted as unnecessary, but I still need it for things to work.  Slowing the clock down before the memcpy fixed all the save/load problems and issues with corrupt strings etc. in my case.  It works great and I'm quite happy with it.

**Reply by frntc** [reply]
That's weird (maybe I missed something), but readConfiguration is called after the Pico has startet, and in this state it's clocked slow enough so that the flash must work.

**Reply by rsquared68** [reply]
Until you wrote that, it didn't occur to me that of course the Pico should start with a normal clock and not have an issue, unless something is configured wrong in my SDK. 

Maybe the actual issue is caused by what looks to me (Pico newbie) like an out-of-place SET_CLOCK_FAST in the writeConfiguration function

```
void writeConfiguration()
{
	SET_CLOCK_125MHZ
	//sleep_ms( 2 );
	DELAY_Nx3p2_CYCLES( 85000 );
	flash_range_erase( FLASH_CONFIG_OFFSET, FLASH_SECTOR_SIZE );

	uint16_t c = crc16( config, 62 );
	config[ 62 ] = c & 255;
	config[ 63 ] = c >> 8;

	flash_range_program( FLASH_CONFIG_OFFSET, config, FLASH_PAGE_SIZE );

	SET_CLOCK_FAST                // <---- why? if you immediately set it to 125MHz again

	SET_CLOCK_125MHZ
	DELAY_Nx3p2_CYCLES( 85000 );
	readConfiguration();
}
```
It's possible that on my HW, the DELAY isn't long enough for the 125MHz to take effect before it starts readConfiguration.  This could happen when writeConfiguration is called from somewhere else, which if I remember happens in the config bus-handling part.  That doesn't explain startup; maybe reset behavior?  I actually only tested the SKPico in a 326298 board, which if you know it has a wonky reset circuit. I modified the board it to be more like the 250407 so EasyFlash 3 will work, but I don't rule out that the Pico could get an extra reset on powerup.   If you like I can do further testing and send screenshots comparing how it looks corrupted/fixed with and without the extra slowdown and other options like removing the extra speed changes or increasing the delay, but it now all works well for me so I've set it aside.

**Reply by Bigi74** [reply]
Hi, I have the same problems with the trash in the configuration menu, everything is mixed up and I can't save the configuration, the version I have in the pico is the latest SKpico2_DAC_PWM_RGB.uf2 on firmware hw ver.0.2 fw0.21 in mono version. I use C128 Pal. Music and RGB work but without any control and it seems to me that it occasionally hiccups, for me menu for prg files works without the trash and I put a one program for testing, after shutting down and restarting the program is in first place and works without problems. I also have Action Replay VI installed but even without it I have the same symptoms, maybe I'm also bothered by the too fast flash on the clone pico?
Can you send me the processed flash so I can try it with it and by the way I'll also check what revision the motherboard is and it seems to me that it has a German keyboard layout.
Sorry for the long post and best regards and awesome project Frntc

**Reply by rsquared68** [reply]
These builds are what I'm using and built with only LED and no RGB; one is PWM-only and the other DAC-only:

[r2-build.zip](https://github.com/user-attachments/files/17758987/r2-build.zip)

Consistent with the above discussion I'm not sure why these work but they do work for me.  If the flash starts behaving better (trash in menu gone, save/load works, PRG menu works) but you have other problems you'll still need to tune the timings with the patcher.  Notably, in my case everything worked except the Flex tune accessed by hitting 'p' in the menu would not play until I adjusted the timings.   In my case judging by my shell history I used  .\SKpicopatch.exe .\SKpico202-DAC-LED-RJR.uf2 10 11

If this works I can rebuild for you with more features enabled.  Note if you also use the RAD Expansion it probably won't properly detect these builds because I think the version string is different.

**Reply by frntc** [reply]
I you really think there is a problem with the flash -- please take a photo of your Pico (clone) so that we can see the flash IC model. 

But: if there are problems launching the menu, then please follow the trouble shooting section and adjust the timings!

**Reply by rsquared68** [reply]
I just got around to upgrading my clone to v0.21 tonight, and I encountered the same problem that I had with the earlier versions similar to what various folks are describing where the flash access causes issues with garbled screens and lockups when using the release build, independent of any timing settings.  Since I had a "solution" from experimenting with the FW earlier versions I just implemented it again and rebuilt from source--and v0.21 now works perfectly.  I totally agree with you that my "fix" of adding another SET_CLOCK_125MHZ at the start of readConfiguration() shouldn't do anything but this time that was all I changed and that fixed it for my build.  I've attached shots of what the bad behavior looks like before the change, and my version of SKpico.c that works for me so you can diff it to see what I did.  It's just strange that this makes it work, but I'm happy with the result.  This is the black YD-RP240 clone, and the SPI flash is marked as a **Winbond 25Q32JVSIQ part with datecode 2135**.  I've scanned the timing settings and for me phi2=10 and read_bus=11 is the middle of the "where it meets timing" matrix, so I probably hardcoded that in the attached source.

![before-fix-prg](https://github.com/user-attachments/assets/48ae6930-e3b6-4f2b-a6b4-ada3b60b0c4a)
![before-fix-config](https://github.com/user-attachments/assets/2405cf30-c440-4526-8438-f7a5dc642e09)

[SKpico.zip](https://github.com/user-attachments/files/18192808/SKpico.zip)

**Reply by Aelius00** [reply]
Hi, I'm still a bit too new at this to recompile the uf2. If you're willing and have time to recompile the DAC_PWM_RGB led enabled with your updated code and make a uf2 file for me. I'd like to give it a go and flash it to my pico. I'd truly appreciate.

**Comment by Bigi74** [comment]
Thanks for flash,i will try and post results. i dont have RAD only  AR
VI and epyx fastload but most of time use arVI
thanks again for helping me.

pet, 15. stu 2024. u 00:30 rsquared68 ***@***.***> napisao je:

**Reply by Bigi74** [reply]
Hi same with your flashes, i solder A5 on cpu,did i need other wire or is that enough? after few reset/shutdown and pico almost fries my speakers with extremly loud amplf,fast shutdown and again playing with reset ans poweroff i manage to get les garbeged menu and stereo sound,omg that sound almost like Amiga i listen half hour music from menu and shutdown c128 i think it will be same problems when i poweron 128, looks like is problem with timmings like frntc says. here some pictures from menu and pico, thanks guys for help.
![WhatsApp Slika 2024-11-15 u 16 21 40_029a8589](https://github.com/user-attachments/assets/e84a85ca-927b-40ab-8139-a1cd6c2c5276)
![WhatsApp Slika 2024-11-15 u 16 21 41_6f98f7d6](https://github.com/user-attachments/assets/f3ff25ef-729b-408c-bd05-ab0922d87394)
![WhatsApp Slika 2024-11-15 u 16 21 41_c3d1e285](https://github.com/user-attachments/assets/b4cb141d-2de2-459d-baf6-1ca77b108ada)
![WhatsApp Slika 2024-11-15 u 16 29 11_67c21488](https://github.com/user-attachments/assets/39c48841-16b5-4b6b-8486-5e4acb601c83)

https://github.com/user-attachments/assets/4b1f8fce-2ed6-4b81-9570-134172b9c069

**Reply by frntc** [reply]
if you'd use the official firmware it would diagnose the timings when launching the menu... this is not 100% failsafe though. Anyways, it looks like there's something wrong with the timings.

**Reply by Bigi74** [reply]
Hi, in pico now is "SKpico2_DAC_RGB.uf2" firmware i try today difrent timings from c64 floppy and found that if 11 replace with 3 i got better menu but not all is perfect on other side i replace 15 with 10 and i got better results, then i become impatient and set timings to    3 and 5 then pico freezes and no sound and cant call menu or use timing prg, tomorow unscrew 128 and reflash :) where is  official firmware to use for auto diagnose Frntc?is one from your package? and btw i ordered 1 more from restore store today
Sorry for bother you Frntc.

**Reply by frntc** [reply]
you can try small values for both timings (if one needs to be smaller, than the other one typically as well). The diag is part of the normal firmware (latest version). As said, it's not absolutely failsafe, but typically displays a warning if something is wrong (at least it did for all the testers)

**Reply by Bigi74** [reply]
Update: I have with c128,2 c64 and board on one have sid socket and after reflash i insert pico in c64 and same problems like on c128  lookslike meybe is cold solder pin, i dont belive that both mbo have timing problem and i notice that right channel most of time is much lover in volume than left if is configure on $d420. if is $d400, right  channel have same strong volume like left from speakers. 
Any hint from your expirience where to start refresh solder pico? or sidkick? frntc?

**Reply by frntc** [reply]
If the menu is not working reliable (despite adjusting timings) then one of the two CBTD3861 (or their soldering) might cause the problem. Both are required for the bus communication so there's no priority which one to test first.

If you configure a 2nd SID at $d420 (or any address other than $d400) then the maximum volume is not reached by a single emulated SID, i.e. single-SID tunes are played with lower volume. This is because when you set both to $d400 you get 2 SIDs playing the same tune.

Which SKpico-PCB do you have and how do you output the sound? (external DAC or built-in?)

**Reply by Bigi74** [reply]
version 0.2 with dac builtin and sound is connected to 2.1 pc speaker

**Comment by Bigi74** [comment]
![sidkick-pico-v02-dac-2024-frenetic](https://github.com/user-attachments/assets/500a919b-b92b-4827-95d7-0ba34bea5bee)

**Reply by frntc** [reply]
Ok, I was asking because with this version you could route the right channel through the mainboard (but that's not recommended, and I assume you're not doing that, correct?)

**Comment by Bigi74** [comment]
Correct.

sri, 20. stu 2024. 22:05 frntc ***@***.***> je napisao:

**Reply by frntc** [reply]
Ok, then you should double and triple check the soldering and that the pins tightly fit into the SID-socket. I think it's unlikely that one of the CBTD3861 is defective.

**Comment by Bigi74** [comment]
Ok,thanks for help 👍🏻

sri, 20. stu 2024. 22:10 frntc ***@***.***> je napisao:

**Comment by Aelius00** [comment]
Hi, I've been searching everywhere for an answer. My pico was working fine but I decided to use Mssiah and add Pin7 for adding $de00. It worked well... However... Now almost every time I go in to the SIDKick menu it's garbled. Even overwrites settings by itself. Hitting F8 sometimes stabilizes it. I unsoldered the pin and reflashed it but the problem still persists. I have no idea what's gone wrong. I think it keeps thinking it's got a conflict with the expansion port. Perhaps the CBTD3861is fried? lol

Maybe the auto detection goes nuts somehow? Hmmpf.

It's the same black board as pictured above. Leds work. It actually worked perfectly before I added $de00.

Edit: I see another post with the same issue. Seems to maybe be a hardware fault? In that case maybe it might be better to flash the pico without a menu but already with the required SID options pre-set in the chosen uf2. Because so far It does seem to retain the SID settings as long as you don't enter the menu.

**Reply by frntc** [reply]
The configuration is overwritten with default once you flash the firmware again. Also cartridges do influence the bus signals, I think you should try different timing values first.

**Reply by Aelius00** [reply]
I've tried many different values. All of them seem to make it worse. Small increments to large. 

Edit: Hmm. Seems to get worse and worse the more I try to enter the menu. Even on default values it now just freezes up the entire computer. Sometimes wont even enter the menu. Sound becomes really nasty too.

**Reply by frntc** [reply]
Can you try a different Pico (clone)? You're using one of the black clones, right? I'm asking because of the build-quality issues...

**Reply by Aelius00** [reply]
Sadly, I don't have a desolder gun anymore at the moment. It died last month. I also don't have any spare SIDkick boards to build a new one. If user [rsquared68](https://github.com/rsquared68) is willing to compile a new uf2 with the updated setting with DAC_PWM_RGB enabled, I'd be willing to give that a go. Perhaps you could compile it? I don't trust my knowledge enough yet not to brick something.

**Reply by frntc** [reply]
Independent of your SKpico variant you can try the firmware he posted above (just to check if the menu comes up). But as @rsquared68 mentioned, I have no idea why that redundant line of code would change anything, if at all, I'd increase the waiting time for the flash which should have the same effect.
